File: Practica 3/cola/src/Cola/Cola.java
package Cola;

class Cola {
    private long[] arreglo;
    private int inicio;
    private int fin;
    private int n;
    private int size;

    public Cola(int n) {
        this.n = n;
        this.arreglo = new long[n];
        this.inicio = 0;
        this.fin = -1;
        this.size = 0;
    }

    public void insert(long e) {
        if (isFull()) {
            System.out.println("Cola llena! No se pudo agregar " + e);
            return;
        }
        fin = (fin + 1) % n;
        arreglo[fin] = e;
        size++;
        System.out.println("Elemento añadido: " + e + ", Cola actual: " + mostrarCola());
    }

    public long remove() {
        if (isEmpty()) {
            System.out.println("Cola vacía! No se puede eliminar.");
            return -1;
        }
        long elemento = arreglo[inicio];
        inicio = (inicio + 1) % n;
        size--;
        System.out.println("Elemento eliminado: " + elemento + ", Cola actual: " + mostrarCola());
        return elemento;
    }

    public long peek() {
        if (isEmpty()) {
            System.out.println("Cola vacía! No hay elementos en la cima.");
            return -1;
        }
        return arreglo[inicio];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == n;
    }

    public int size() {
        return size;
    }

    private String mostrarCola() {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < size; i++) {
            sb.append(arreglo[(inicio + i) % n]);
            if (i < size - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        Cola cola = new Cola(5);
        int[] elementos = {10, 20, 30, 40, 50, 60};

        for (int num : elementos) {
            cola.insert(num);
        }

        System.out.println("Elemento eliminado: " + cola.remove());
        System.out.println("Elemento en la cima: " + cola.peek());
    }
}

File: Practica 3/cola/Cola.py
class Cola:
    def __init__(self, n: int):
        self.__n = n
        self.__arreglo = [0] * n
        self.__inicio = 0
        self.__fin = -1
        self.__size = 0

    def insert(self, e: int):
        if self.is_full():
            print(f"Cola llena! No se pudo agregar {e}")
            return
        self.__fin = (self.__fin + 1) % self.__n
        self.__arreglo[self.__fin] = e
        self.__size += 1
        print(f"Elemento añadido: {e}, Cola actual: {self.__arreglo[self.__inicio:self.__fin + 1]}")

    def remove(self) -> int:
        if self.is_empty():
            print("Cola vacía! No se puede eliminar.")
            return -1
        elemento = self.__arreglo[self.__inicio]
        self.__inicio = (self.__inicio + 1) % self.__n
        self.__size -= 1
        print(f"Elemento eliminado: {elemento}, Cola actual: {self.__arreglo[self.__inicio:self.__fin + 1]}")
        return elemento

    def peek(self) -> int:
        if self.is_empty():
            print("Cola vacía! No hay elementos en la cima.")
            return -1
        return self.__arreglo[self.__inicio]

    def is_empty(self) -> bool:
        return self.__size == 0

    def is_full(self) -> bool:
        return self.__size == self.__n

    def size(self) -> int:
        return self.__size


if __name__ == "__main__":
    cola = Cola(5)
    for num in [10, 20, 30, 40, 50, 60]:
        cola.insert(num)

    print(f"Elemento eliminado: {cola.remove()}")
    print(f"Elemento en la cima: {cola.peek()}")
